rust/gateway/src/eventloop.rs

```diff
@@ -228,34 +157,94 @@ impl Eventloop {
+            req.client.id,
+            req.client.peer.preshared_key,
+            req.client.payload.ice_parameters,
+            PublicKey::from(req.client.peer.public_key.0),
+            ips,
+            req.relays,
+            req.client.payload.domain,
+            req.expires_at,
+            req.resource.into_resolved(addresses),
+        ) {
+            Ok(accepted) => {
+                self.portal.send(
+                    PHOENIX_TOPIC,
+                    EgressMessages::ConnectionReady(ConnectionReady {
+                        reference: req.reference,
+                        gateway_payload: GatewayResponse::ConnectionAccepted(accepted),
+                    }),
                 );
-                bail!("Protocol error: Request for DNS resource without the subdomain being tried to access.")
-            };
-
-            let addresses =
-                tokio::task::spawn_blocking(move || resolve_addresses(&domain.to_string()))
-                    .await??;
-
-            Ok(ResourceDescription::Dns(ResolvedResourceDescriptionDns {
-                id: dns.id,
-                domain: dns.address,
-                name: dns.name,
-                addresses,
-            }))
+
+                // TODO: If outbound request fails, cleanup connection.
+            }
+            Err(e) => {
+                let client = req.client.id;
+
+                self.tunnel.cleanup_connection(&client);
+                tracing::debug!(%client, "Connection request failed: {:#}", anyhow::Error::new(e));
+            }
+        }
+    }
+
+    pub fn allow_access(&mut self, result: Result<Vec<IpNetwork>, Timeout>, req: AllowAccess) {
+        let addresses = result
+            .inspect_err(|e| tracing::debug!(client = %req.client_id, reference = %req.reference, "DNS resolution timed out as part of allow access request: {e}"))
+            .unwrap_or_default();
+
+        let maybe_domain_response = self.tunnel.allow_access(
+            req.resource.into_resolved(addresses),
+            req.client_id,
+            req.expires_at,
+            req.payload,
+        );
+
+        if let Some(domain_response) = maybe_domain_response {
+            self.portal.send(
+                PHOENIX_TOPIC,
+                EgressMessages::ConnectionReady(ConnectionReady {
+                    reference: req.reference,
+                    gateway_payload: GatewayResponse::ResourceAccepted(ResourceAccepted {
+                        domain_response,
+                    }),
+                }),
+            );
+        }
+    }
+}
+
+async fn resolve(domain: Option<Dname>) -> Vec<IpNetwork> {
+    let Some(domain) = domain.clone() else {
+        return vec![];
+    };
+
+    let dname = domain.to_string();
+
+    match tokio::task::spawn_blocking(move || resolve_addresses(&dname)).await {
+        Ok(Ok(addresses)) => addresses,
+        Ok(Err(e)) => {
+            tracing::warn!("Failed to resolve '{domain}': {e}");
+
+            vec![]
```

Comment:
We should probably have a dedicated message for this, but for now maybe we need to handle it from the client-side disconnecting from the gateway.

This will be left in a weird state connected to a gateway which might have no resources associated.

Reply:
It could be superious? It prompted me to write up https://github.com/firezone/firezone/issues/4267.